src/MongoDB.Driver/Linq/Processors/ExpressionValidator.cs

```diff
@@ -0,0 +1,122 @@
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace MongoDB.Driver.Linq.Processors
+{
+    /// <summary>
+    /// The expression visitor to validate a LINQ expression.
+    /// </summary>
+    internal sealed class ExpressionValidator : ExpressionVisitor
+    {
+        private readonly List<string> _topLevelInParameters = new List<string>();
+        private List<string> _inParametersScope = new List<string>();
+
+        /// <summary>
+        /// Validates the input expression.
+        /// </summary>
+        /// <param name="expression">The expression.</param>
+        public static void Validate(Expression expression)
+        {
+            var validator = new ExpressionValidator();
+
+            var lambdaExpression = expression as LambdaExpression;
+            if (lambdaExpression?.Parameters != null && lambdaExpression.Parameters.Any())
+            {
+                validator._topLevelInParameters.AddRange(lambdaExpression.Parameters.Select(c => c.Name));
+            }
+
+            validator.Visit(expression);
+        }
+
+        /// <inheritdoc />
+        protected override Expression VisitMethodCall(MethodCallExpression node)
+        {
+            if (node.Arguments.Any())
+            {
+                foreach (var argument in node.Arguments.Where(c => !(c is ConstantExpression || c is ParameterExpression))) // do not check the bottom of tree branches.
+                {
+                    var usedParameters = ParameterExpressionCollector.Collect(argument).ToList();
+
+                    if (!usedParameters.Any())
+                    {
+                        continue;
+                    }
+
+                    var memberExpression = argument as MemberExpression;
+                    if (memberExpression != null)
+                    {
+                        // add top-level expression parameters to the list
+                        // with parent's parameters.
+                        // NOTE: these parameters will be added to all members,
+                        // even if the member is not in the top-level, but since
+                        // top-level parameters should not be placed in any level 
+                        // except of direct children, it's not a problem
+                        AddTopLevelExpressionParameters(_inParametersScope);
+                        continue;
```

Comment:
Why is this only for MemberExpression?

I don't mean to imply this is wrong... I'm just asking for an explanation.

Reply:
This is the same skipping logic as for `ConstantExpression` and `ParameterExpression`.
But `MemberExpression` is also a left branch(arguments[0]) of MethodCallExpression (which means that the right branch of this `MethodCallExpression` will be one level down for current expression) - so in this case, we additionally need to compensate the top-level params for the right branch of current `MethodExpression`.